### PR TITLE
Make slides to work with latest Reveal

### LINKED ISSUE #34: slides display problem: reveal.js version conflict?
I'm using IPython 3.0.0 and having various issues with slide conversion.
(I had to manually install mistune (pip installed  mistune.__version__ '0.5.1')

Scenario 0:
`ipython nbconvert --to slides TestSlides.ipynb --post serve`
This works, but you see it is getting an old reveal 2.6.2
`[NbConvertApp] Redirecting reveal.js requests to https://cdn.jsdelivr.net/reveal.js/2.6.2`

Note, if you use nbviewer it uses reveal 3.1.0

The problem is that you can't put these on a live server unless it is has reveal.js installed in that local directory.

Scenario 1:
```
ipython nbconvert --to slides TestSlides.ipynb
python -m SimpleHTTPServer 8000
````
This browser sees the html, but they don't act as slides b/c reveal.js isn't found. 

Scenario 2:
```
ipython nbconvert --to slides TestSlides.ipynb
wget https://github.com/hakimel/reveal.js/archive/3.0.0.tar.gz
tar -xvf 3.0.0.tar.gz
mv reveal.js-3.0.0 reveal.js
python -m SimpleHTTPServer 8000
```
Now the browser sees the reveal.js (js and css) resources, but for some reason it renders as a blank screen. It also still has problem that you can't deploy without making a copy of reveal.js in the corresponding directory on the live server.

Scenario 3:
```
ipython nbconvert --to slides TestSlides.ipynb  --reveal-prefix https://cdn.jsdelivr.net/reveal.js/2.6.2
python -m SimpleHTTPServer 8000
```

This works now locally and on live server, but with old reveal.js

Scenario 4:
```
ipython nbconvert --to slides TestSlides.ipynb  --reveal-prefix http://cdn.bootcss.com/reveal.js/3.1.0
python -m SimpleHTTPServer 8000
```
This gives similar behavior as Scenario 2, the resources are there but the browser renders an empty screen. In both Scenario 2 & 4 I see these errors in the console:
```
[Error] ReferenceError: Can't find variable: Reveal
	global code (TestSlides.slides.html, line 333)
[Error] Error: Mismatched anonymous define() module: function () {
			root.Reveal = factory();
			return root.Reveal;
		}
http://requirejs.org/docs/errors.html#mismatch
	da (require.min.js, line 8)
	w (require.min.js, line 12)
	M (require.min.js, line 15)
	(anonymous function) (require.min.js, line 26)
```

### PR DESCRIPTION
**Pull request by damianavila**
Closes #34 

* [x] Change how to load latest reveal
* [x] Update CDN to get the latest `reveal.js`
* [x] Load the mathjax macro in a proper place at `head`

### CODE REVIEW

**Comment by damianavila** [comment]
WIP yet:

* [ ] Make scroll work again
* [ ] Render math on fragments too

**Comment by damianavila** [comment]
@bollwyvl probably this is interesting for you. I migrate the `overflow-y`into the `reveal` class. In this way we only have a functional scrolling bar when there is content exceeding the slides view. I chose to use `scroll` instead of `auto`, because `auto` generates an horizontal movement when you switch from a "non scrolling" slide to a "scrolling" one. We discussed this in the past in the nbviewer repo and I think this solution, even it is not ideal, it solves some of the problem on mobile and actually represent better the "scroll" status of a slide...

**Comment by bollwyvl** [comment]
Very cool, looks much cleaner!

**Comment by damianavila** [comment]
ping @jupyter/nbconvert

File: nbconvert/templates/html/slides_reveal.tpl [hunk 1]
```diff
@@ -168,37 +171,52 @@ a.anchor-link {
...
-]
-});
-</script>
-
-<!-- Loading mathjax macro -->
-{{ mathjax() }}
+require(
+    {
+      // for really large notebook this probably makes sense
+      waitSeconds: 30
```
Comment (takluyver) [inline nbconvert/templates/html/slides_reveal.tpl]: What is this timeout for? The size of the notebook shouldn't affect how long it takes to load JS libraries, right?

File: nbconvert/templates/html/slides_reveal.tpl [hunk 2]
```diff
@@ -168,37 +171,52 @@ a.anchor-link {
...
+                { src: "{{resources.reveal.url_prefix}}/lib/js/classList.js",
+                  condition: function() { return !document.body.classList; } },
+                { src: "{{resources.reveal.url_prefix}}/plugin/notes/notes.js",
+                  async: true,
+                  condition: function() { return !!document.body.classList; } }
+            ]
+        });
+
+        var update = function(event){
+          window.scrollTo(0,0);
```
Comment (takluyver) [inline nbconvert/templates/html/slides_reveal.tpl]: Is the scrollTo needed here? It seems unrelated to rendering mathjax.

File: nbconvert/templates/html/slides_reveal.tpl [hunk 3]
```diff
@@ -74,6 +74,9 @@ if( window.location.search.match( /print-pdf/gi ) ) {
 <script src="{{resources.reveal.url_prefix}}/lib/js/html5shiv.js"></script>
 <![endif]-->
 
+<!-- Loading the mathjax macro -->
+{{ mathjax() }}
```
Comment (takluyver) [inline nbconvert/templates/html/slides_reveal.tpl]: Where is this macro defined? I can't immediately find it in our code.

**Comment by takluyver** [comment]
Overall I think this looks good, there are a few specific bits I've asked for clarification on.

File: nbconvert/templates/html/slides_reveal.tpl (hunk 1 above)
Comment (damianavila) [inline nbconvert/templates/html/slides_reveal.tpl]: You are right... I was working on RISE at the same time, where the time to load the live notebook could affect the reveal.js load and arise a timeout...
Certainly, I don't think this setting is needed here, unless I am missing something else... so I will delete it for now...

File: nbconvert/templates/html/slides_reveal.tpl (hunk 2 above)
Comment (damianavila) [inline nbconvert/templates/html/slides_reveal.tpl]: It is not related with mathjax, I should add a comment here... also I have to check if the scroll is still working after change the overflow into the reveal class... essentially you want to scroll to the top on every `slidechange` event (otherwise when you come back to a previous slide, you probably miss content because the slide keep scrolled to the bottom), so probably I have to rewrite this to only scroll on the `slidechange` event and not in the fragments events.

File: nbconvert/templates/html/slides_reveal.tpl (hunk 3 above)
Comment (damianavila) [inline nbconvert/templates/html/slides_reveal.tpl]: At `nbconvert/nbconvert/templates/html/`, next to others html templates.

Comment (takluyver) [inline nbconvert/templates/html/slides_reveal.tpl]: Aha, thanks.

File: nbconvert/templates/html/slides_reveal.tpl (hunk 1 above)
Comment (takluyver) [inline nbconvert/templates/html/slides_reveal.tpl]: OK, if you want to delete this and add the comment below, I think this is good to go.

**Comment by damianavila** [comment]
@takluyver, made some additional changes:

1. I would like to use the waitSecond settting because it makes sense when you are in slow connections loading reveal from a CDN. I set that t 15 seconds.
2. I fix the scroll to top on the slidechange event. Now the code is self-explanatory and decupled from the mathjax update, so there is no need for a comment, I think...

@bollwyvl, regarding my last comment about the scrolling, regrettably I still see the overflow bar in some no-scroll slides because it seems to take next largest slide to set up the bars... but still is better than before because people not using any "scrolling" slides will have a no-scrolling slideshow (with the bar but deactivated).

**Comment by bollwyvl** [comment]
Bummer about the scrolling. I hate to be defending slides as a medium, but
I just have trouble with the "user has so much crap to say it doesn't fit
on a weirdly-margined-slide use case" :P

Apropos of https://github.com/jupyter/nbviewer/issues/484 (bokeh table
css), is there a reason we don't have a #notebook somewhere in slides?

Also, might it be possible to break up the reveal stuff into even more
blocks to make nbviewer easier? I had to strip most of the page chrome to
get all the paths right, which lead to the real bug in that issue. If I
could sub the template and pull the no-fooling `<div class="reveal">` that
might help avoid these kinds of problems.

**Comment by damianavila** [comment]
>Bummer about the scrolling. I hate to be defending slides as a medium, but
I just have trouble with the "user has so much crap to say it doesn't fit
on a weirdly-margined-slide use case" :P

Can you elaborate a little more on this...

>Apropos of https://github.com/jupyter/nbviewer/issues/484 (bokeh table
css), is there a reason we don't have a #notebook somewhere in slides?

No reason I am aware of... or I remember...

>Also, might it be possible to break up the reveal stuff into even more
blocks to make nbviewer easier? I had to strip most of the page chrome to
get all the paths right, which lead to the real bug in that issue. If I
could sub the template and pull the no-fooling `<div class="reveal">` that
might help avoid these kinds of problems.

Yes, please, open an issue for that... there is another asking for that to inject reveal configure in a custom template, I think it makes sense to decouple things with more block to help people (and us) to use it in custom settings (like nbviewer).

**Comment by bollwyvl** [comment]
> Can you elaborate a little more on this...

I think the phrase I used before was that they "break the medium"... slides "unintentionally" bigger (i.e. not infinite like prezi, etc.) than the screen...
- are not supported on any other slide platform
- magnify the "i'm going to read you this slide" presentation style... most people can read well north of 300 words per minute in their mother tongue, and showing someone a dense slide, they'll likely tune you out until they've read it
- break URLs: no way to link to a particular scrolled location. not that the reveal ones are great to begin with
- are not liked by reveal, hence all our troubles
  - will likely never printable or exportable, one of the use cases i value a lot but haven't been able to handle succinctly... even without scrolling slides... but at least it is theoretically doable!
- which require a scroll wheel/bar during a talk is distracting to the speaker and the listeners, and content will be in a half-revealed state, which again breaks focus for both parties
  - if the content really is big, like a big diagram or visualization, then temporary full screen/scroll/zoom on that content is a better fit... with text and margins so big, it's hard to even get 80 character code on a screen
- break touch gestures (unless you've fixed that!)

perhaps an explicit `longslide` slidetype that, if found, would enable the scrolling, while `slide` would conform to the reveal model would be less surprising.

> No reason I am aware of... or I remember...

wasn't sure if there was some un-reset stuff from notebook that would mess us up... in browser testing suggests it looks okay...

> Yes, please, open an issue for that...

Will do!

> inject reveal configure in a custom template

Would love to see this in the notebook metadata as the "last word"... but i could see how having nbviewer options, config options (as per rise) etc.

**Comment by damianavila** [comment]
>perhaps an explicit longslide slidetype that, if found, would enable the scrolling, while slide would conform to the reveal model would be less surprising.

Well, lot of good reasons, we should discuss this point again... I like the idea of a long slide, but no sure how complicated could be... I passed those critics to new issue to discuss further: #78

>wasn't sure if there was some un-reset stuff from notebook that would mess us up... in browser testing suggests it looks okay...

Great, should be an easy fix...

>Will do!

Thanks

>Would love to see this in the notebook metadata as the "last word"...

Yep... I agree, we should probably re-discuss that point, or provide a plugin (extension) to configure this from the notebook metadata... maybe you can open a new issue to discuss that :wink:

**Comment by damianavila** [comment]
btw, @takluyver I think this is ready... all the other discussion are in parallel and to be solved later.

**Comment by takluyver** [comment]
Thanks, I think this looks good.
